python/ray/tune/trial_runner.py

```diff
@@ -225,6 +232,7 @@ def has_resources(self, resources):
 
     def _get_next_trial(self):
         self._update_avail_resources()
+        self._update_trial_queue()
```

Comment:
Is it possible for the algorithm to generate NOT_READY, while the entire tuning loop exits?

Reply:
Good point - I guess the main confusion here stems from the naming of `NOT_READY`. Perhaps it should be named as `PASS`? I just pushed a commit that does this.

`PASS` should only be returned when it wants the TrialRunner to continue executing without fetching a new trial. 

If it does NOT want the TrialRunner to continue executing, it can block until the algorithm finishes calculating whatever it needs to calculate. 

There are a few cases in mind:
1. When the algorithm is throttled by maximum concurrency
2. When the algorithm is throttled by some internal calculation. In this case, you'll either need to block per calculation or you want to let the calculation run on a separate thread.

If the SearchAlgorithm wants to flip between blocking and passing, then it could do that - after all, it knows the trials that it generates and it also knows the trials that complete/error/get killed.

Reply:
Thanks for the clarification.

If an algorithm returns PASS at the moment, under what circumstance will it get another chance to `try_suggest`?  Currently, it seems that the runner will exit as long as all trials complete. I think we may need a terminating flag such as `END` to tell the runner explicitly that there is no more trials later for this algorithm.

Reply:
Done - I made it so that the event loop driven by the TrialRunner was mainly unaffected. 

I added a `is_finished` to the SearchAlgorithm class along with a blocking switch on `update_trial_queue`, which blocks and polls if all trials are done but the search algorithm is still not done (with a timeout). `SearchAlgorithm.is_finished` should provide the "terminating flag" as requested. 

Let me know what you think.